### PR TITLE
FINALLY Feature/notifying

### PR DESCRIPTION
**Pull request by Minji1234**
FINALLY I succeeded to get alarm.
Let's delete chrome and move to firefox. Bad chrome..
I made NewAnswer.js and relevant codes both in frontend and backend.
Thank you for helping me to deal with annoying chrome problem!!!!!!

### CODE REVIEW

File: oga/backend/users/signals.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 """signals and appropriate actions"""
 from django.db.models.signals import post_save
 from django.dispatch import receiver
-from .models import User, Profile
+from .models import User, Profile, Question, Answer, Location
```
Comment (taehwoi) [inline oga/backend/users/signals.py]: pylint complains about unused imports

File: oga/backend/users/signals.py [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
+from .utils.send_push import send_push
+from .utils.haversine import distance, haversine
```
Comment (taehwoi) [inline oga/backend/users/signals.py]: pylint complains about unused imports

File: oga/backend/users/utils/haversine.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    Calculate the great circle distance between two points 
+    on the earth (specified in decimal degrees)
+    """
+    # convert decimal degrees to radians 
+    lon1, lat1, lon2, lat2 = map(radians, [lon1, lat1, lon2, lat2])
+
+    # haversine formula 
+    dlon = lon2 - lon1 
+    dlat = lat2 - lat1 
+    a = sin(dlat/2)**2 + cos(lat1) * cos(lat2) * sin(dlon/2)**2
```
Comment (taehwoi) [inline oga/backend/users/utils/haversine.py]: pylint complains about single character variables

File: oga/backend/users/utils/haversine.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+from math import radians, sin, cos, asin, sqrt
+from users.models import User, Profile, Question, Answer, Location
```
Comment (taehwoi) [inline oga/backend/users/utils/haversine.py]: Unused imports

File: oga/backend/users/views/views_answer.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    answer_author = get_user(request)
+    author = Profile.objects.get(user=answer_author)
+    answer_content = req_data['answer_content']
+
+    answer = Answer(question_id=question_id,
+                    author=answer_author.id,
+                    question_type=question_type,
+                    answer_content=answer_content)
+    answer.save()
+    return JsonResponse(response_dict, status=200)
```
Comment (taehwoi) [inline oga/backend/users/views/views_answer.py]: Where is response_dict created in this function? I can't find it.

File: oga/backend/users/views/views_question.py [hunk 6]
```diff
@@ -16,15 +16,27 @@ def questions(request):
     """questions api"""
     req_data = json.loads(request.body.decode())
     location = req_data['target_location']
-    content = req_data['content']
+    question_type = req_data['content']
```
Comment (taehwoi) [inline oga/backend/users/views/views_question.py]: why is content named question_type?

File: oga/backend/users/views/views_question.py [hunk 7]
```diff
@@ -16,15 +16,27 @@ def questions(request):
...
     location, _ = Location.objects.get_or_create(name=location['name'],
                                                  latitude=location['latitude'],
                                                  longitude=location['longitude'])
 
-    question = Question(author=user, location_id=location,
-                        content=content)
+    question = Question(author=user, content=question_type, location_id=location)
     question.save()
-    response_dict = {'id': question.id}
+    response_dict = {'id': question.id, 'author_id': user.id, 'content': question.content, 'target_location': question.location_id.id}
```
Comment (taehwoi) [inline oga/backend/users/views/views_question.py]: shouldn't target_location just be question.location_id?

Comment (taehwoi) [reply oga/backend/users/views/views_question.py]: And pylint complains about long lines

File: oga/backend/users/views/views_question.py [hunk 8]
```diff
@@ -16,15 +16,27 @@ def questions(request):
...
+
+@check_login_required
+@check_request
+@require_http_methods(["GET"])
+@csrf_exempt
+def get_question(request, question_id):
+    """question api"""
+    req_data = json.loads(request.body.decode())
+    question = Question.objects.get(id=question_id)
+    response_dict = {'id': question.id,
```
Comment (taehwoi) [inline oga/backend/users/views/views_question.py]: I think the frontend needs the location name of the questions as well

File: oga/frontend/src/containers/Answer/NewAnswer.js [hunk 9]
```diff
@@ -0,0 +1,131 @@
+import React, { Component } from 'react';
+import './NewAnswer.css';
+
+import { connect } from 'react-redux';
+import * as actionCreators from '../../store/actions/';
+
+import { Redirect } from 'react-router-dom';
+import { push } from 'connected-react-router';
+import Map from '../Map/GoogleMap';
+import PushNotification from '../../components/PushNotification/PushNotification'
```
Comment (taehwoi) [inline oga/frontend/src/containers/Answer/NewAnswer.js:10]: Why is PushNotification imported in this file?

File: oga/frontend/src/containers/Answer/NewAnswer.js [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+class NewAnswer extends Component {
+  state = {
+    answer_content: '....',
+  }
+
+  componentDidMount() {
+    this.props.onGetQuestion(this.props.match.params.id);
+  }
+
+  postQuestionHandler = () => {
```
Comment (taehwoi) [inline oga/frontend/src/containers/Answer/NewAnswer.js]: are we posting question here? or answer?

File: oga/frontend/src/containers/Answer/NewAnswer.js [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    if (this.state.answer_content !== '' && this.props.selectedQuestion.content)
```
Comment (taehwoi) [inline oga/frontend/src/containers/Answer/NewAnswer.js]: I don't understand this condition. Can you please explain? If we got the question from backend, shouldn't if have content?

File: oga/frontend/src/containers/Answer/NewAnswer.js [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+  render() {
+    let place_name = "...";
+    if (this.props.target_location)
+    {
+      place_name = this.props.target_location.name;
+    }
+    let show = null;
+    var qs_type = this.props.selectedQuestion.content;
+    if (qs_type == "LINE") {
```
Comment (taehwoi) [inline oga/frontend/src/containers/Answer/NewAnswer.js]: I think there must be a better way to handle this.
For example, a const table that returns possible answers lists from qs_type?

File: oga/frontend/src/containers/QuestionList/NewQuestion/NewQuestion.js [hunk 13]
```diff
@@ -95,8 +95,9 @@ const mapStateToProps = state => {
 
 const mapDispatchToProps = dispatch => {
   return {
-    createQuestion: (type, content, target_location) =>
-      dispatch(actionCreators.createQuestion({ author: "HI", type: type, content: content, target_location: target_location})),
+    createQuestion: (type, target_location) =>
```
Comment (taehwoi) [inline oga/frontend/src/containers/QuestionList/NewQuestion/NewQuestion.js]: Again, are we creating answers or questions?

File: oga/frontend/src/store/actions/answerActions.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+axios.defaults.xsrfCookieName="csrftoken";
+axios.defaults.xsrfHeaderName="X-CSRFTOKEN";
+
+export const createAnswer_ = (answer) => {
+  return {
+    type: actionTypes.CREATE_ANSWER,
+    question_id: answer.question_id,
+    author_id: answer.author_id,
+    question_type: answer.question_type,
+    answer_content: answer.answer_content,
```
Comment (taehwoi) [inline oga/frontend/src/store/actions/answerActions.js]: can't we just use content, instead of answer_content. I think it's just redundant.

**Comment by taehioum** [comment]
Also, I think it would be nice if features can be seperated. This branch currently have both answering and notifying feature, which makes testing harder.

File: oga/frontend/src/containers/Answer/NewAnswer.js (hunk 10 above)
Comment (Minji1234) [reply oga/frontend/src/containers/Answer/NewAnswer.js]: I forgot to change name. thanks

File: oga/frontend/src/containers/QuestionList/NewQuestion/NewQuestion.js (hunk 13 above)
Comment (Minji1234) [reply oga/frontend/src/containers/QuestionList/NewQuestion/NewQuestion.js]: I think this page is for making question, as it's name is NewQuestion.js

File: oga/frontend/src/containers/Answer/NewAnswer.js (hunk 12 above)
Comment (Minji1234) [reply oga/frontend/src/containers/Answer/NewAnswer.js]: That is good idea. I will make it.

File: oga/frontend/src/containers/Answer/NewAnswer.js (hunk 11 above)
Comment (Minji1234) [reply oga/frontend/src/containers/Answer/NewAnswer.js]: First I changed it to 
postAnswerHandler = () => {
    if (this.state.answered && this.props.selectedQuestion.content)

it means when the user clicked on one of the answer which is specific to a question and the user normally came to answer page by clicking on reply button on main page, and clicked submit button on this page, then that answer will be posted. I meant this.

Comment (Minji1234) [reply oga/frontend/src/containers/Answer/NewAnswer.js]: the selected_question will be set by using id from url.

Comment (Minji1234) [reply oga/frontend/src/containers/Answer/NewAnswer.js]: and Question have question_type whose default is "LINE"

**Comment by taehwoi** [comment]
Great job! works like magic

**Pull request merged**